### PR TITLE
feat: Importance sampling trick

### PR DESCRIPTION
**Pull request by yfw**
# What does this PR do ?

Adds `importance_sampling_enabled` flag to multiple loss by importance weights when generation policy is off-policy.

# Issues
Closes https://github.com/NVIDIA/reinforcer/issues/54

# Usage
* **You can potentially add a usage example below**

```python
# Add a code snippet demonstrating how to use this 
```

# Before your PR is "Ready for review"
**Pre checks**:
- [ ] Make sure you read and followed [Contributor guidelines](/NVIDIA/reinforcer/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you run the unit tests and functional tests locally? Visit our [Testing Guide](/NVIDIA/reinforcer/blob/main/docs/testing.md) for how to run tests
- [ ] Did you add or update any necessary documentation? Visit our [Document Development Guide](/NVIDIA/reinforcer/blob/main/docs/documentation.md) for how to write, build and test the docs.

# Additional Information
* ...

### CODE REVIEW

File: examples/configs/grpo_math_1B.yaml [hunk 1]
```diff
@@ -14,6 +14,7 @@ loss_fn:
   reference_policy_kl_penalty: 0.01
   ratio_eps_min: 0.2
   ratio_eps_max: 0.2
+  importance_sampling_enabled: false
```
Comment (parthchadha) [inline examples/configs/grpo_math_1B.yaml]:
```suggestion
  use_importance_sampling_correction: false
```

File: nemo_reinforcer/algorithms/loss_functions.py [hunk 2]
```diff
@@ -67,13 +67,18 @@ class ClippedPGLossFn(LossFunction):
 
     For REINFORCE/RLOO (when disable_ppo_ratio=True), the formula simplifies to:
     L(θ) = E_t [ π_θ(a_t|s_t) * A_t ] - β * KL(π_θ || π_ref)
+
+    If the generation policy π_θ_gen is off policy, we can enable importance sampling by setting importance_sampling_enabled=True.
+    This multiplies the loss by the importance weights:
+    importance_weights_t = π_θ_old(a_t|s_t) / π_θ_gen(a_t|s_t)
```
Comment (parthchadha) [inline nemo_reinforcer/algorithms/loss_functions.py]: We should add explanation in docs about how/why importance sampling correction is required.

File: nemo_reinforcer/algorithms/loss_functions.py [hunk 3]
```diff
@@ -67,13 +67,18 @@ class ClippedPGLossFn(LossFunction):
...
     """
 
     def __init__(self, cfg: ClippedPGLossConfig):
         self.ratio_eps_min = cfg["ratio_eps_min"]
         self.ratio_eps_max = cfg["ratio_eps_max"]
         self.reference_policy_kl_penalty = cfg["reference_policy_kl_penalty"]
         self.disable_ppo_ratio = cfg.get("disable_ppo_ratio", False)
+        self.importance_sampling_enabled = cfg["importance_sampling_enabled"]
```
Comment (SahilJain314) [inline nemo_reinforcer/algorithms/loss_functions.py]: Update here as well from config name update

File: tests/unit/algorithms/test_loss_functions.py [hunk 4]
```diff
@@ -386,3 +391,88 @@ def test_clipped_pg_loss_zero_mask():
...
+    )  # [-0.5*0.6065, 1.0*1.6487, -2.4*0.8187] = [-0.30325, 1.6487, -1.96488]
+    expected_actor_loss = torch.mean(importance_weighted_max_loss)  # -0.2065
+
+    # KL Loss Calculation
+    r = (
+        ref_lp_masked - curr_lp_masked
+    )  # [-1.0 - (-1.69), -1.0 - (-1.0), -1.0 - (-0.59)] = [0.69, 0.0, -0.41]
+    kl_term_per_token = (
+        torch.exp(r) - r - 1
+    )  # [exp(0.69)-0.69-1, exp(0)-0-1, exp(-0.41)-(-0.41)-1] = [0.3037, 0.0, 0.0737]
```
Comment (SahilJain314) [inline tests/unit/algorithms/test_loss_functions.py]: if the hand calculations here are claude-written, please actually verify with a calculator (I'm guilty of this too😅)

Comment (terrykong) [reply tests/unit/algorithms/test_loss_functions.py]: maybe just convert all of these to assertions? i do think it's quite nice to have the math worked out

File: examples/configs/grpo_math_1B.yaml (hunk 1 above)
Comment (yfw) [reply examples/configs/grpo_math_1B.yaml]: Updated

File: nemo_reinforcer/algorithms/loss_functions.py (hunk 2 above)
Comment (yfw) [reply nemo_reinforcer/algorithms/loss_functions.py]: Added docs about the loss and importance sampling here: https://github.com/NVIDIA/reinforcer/blob/yifu/importance_sampling/docs/guides/grpo.md#loss

File: nemo_reinforcer/algorithms/loss_functions.py (hunk 3 above)
Comment (yfw) [reply nemo_reinforcer/algorithms/loss_functions.py]: Updated

File: tests/unit/algorithms/test_loss_functions.py (hunk 4 above)
Comment (yfw) [reply tests/unit/algorithms/test_loss_functions.py]: Yes, I've checked these manually. Claude actually got many calculations wrong initially 😱 . I've also converted the comments to assertions.

File: docs/guides/grpo.md [hunk 5]
```diff
@@ -108,3 +109,56 @@ This Policy object holds a [RayWorkerGroup](../../nemo_reinforcer/distributed/wo
...
+\end{align*}
+$$
+
+To enable the on-policy KL approximation, set the config `use_on_policy_kl_approximation=True` in the `ClippedPGLossConfig`.
+
+
+#### Importance Sampling Correction
+The policy we use to draw samples, $\pi_{\theta_{\text{old}}}$, is used in both the inference framework and the training framework. To account for this distinction, we refer to the inference framework policy as $\pi_{\text{inference}}$ and the training framework policy as $\pi_{\text{training}}$. As noted in [Adding New Models](../adding_new_models.md#understanding-discrepancies-between-backends), it is possible that the token probabilities from $\pi_{\text{training}}$ and $\pi_{\text{inference}}$ to have discrepancies, leading to off-policy samples. We can correct for this by introducing importance weights between $\pi_{\text{training}}$ and $\pi_{\text{inference}}$ to the first term of the loss function. Let $f_\theta(x) = \min \Big(\frac{\pi_\theta(x)}{\pi_{\theta_{\text{old}}}(x)}A_t, \text{clip} \big( \frac{\pi_\theta(x)}{\pi_{\theta_{\text{old}}}(x)}, 1 - \varepsilon, 1 + \varepsilon \big) A_t \Big)$ represent the first term of loss function. Then,
```
Comment (SahilJain314) [inline docs/guides/grpo.md]:
```suggestion
The policy we use to draw samples, $\pi_{\theta_{\text{old}}}$, is used in both the inference framework and the training framework. To account for this distinction, we refer to the inference framework policy as $\pi_{\text{inference}}$ and the training framework policy as $\pi_{\text{training}}$. As noted in [Adding New Models](../adding_new_models.md#understanding-discrepancies-between-backends), it is possible for the token probabilities from $\pi_{\text{training}}$ and $\pi_{\text{inference}}$ to have discrepancies (from numerics, precision differences, bugs, etc.), leading to off-policy samples. We can correct for this by introducing importance weights between $\pi_{\text{training}}$ and $\pi_{\text{inference}}$ to the first term of the loss function. 

Let $f_\theta(x) = \min \Big(\frac{\pi_\theta(x)}{\pi_{\theta_{\text{old}}}(x)}A_t, \text{clip} \big( \frac{\pi_\theta(x)}{\pi_{\theta_{\text{old}}}(x)}, 1 - \varepsilon, 1 + \varepsilon \big) A_t \Big)$ represent the first term of loss function. Then,
```

File: docs/guides/grpo.md [hunk 6]
```diff
@@ -108,3 +109,56 @@ This Policy object holds a [RayWorkerGroup](../../nemo_reinforcer/distributed/wo
...
+We use the [ClippedPGLossFn](../../nemo_reinforcer/algorithms/loss_functions.py) to calculate the loss for GRPO. Formally,
+
+$$
+L(\theta) = E_{x \sim \pi_{\theta_{\text{old}}}} \Big[ \min \Big(\frac{\pi_\theta(x)}{\pi_{\theta_{\text{old}}}(x)}A_t, \text{clip} \big( \frac{\pi_\theta(x)}{\pi_{\theta_{\text{old}}}(x)}, 1 - \varepsilon, 1 + \varepsilon \big) A_t \Big) \Big] - \beta D_{\text{KL}} (\pi_\theta \| \pi_\text{ref})
+$$
+
+where:
+
+- $\pi_\theta$ is the policy model we are currently optimizing
+- $\pi_{\theta_{\text{old}}}$ is the previous policy model
```
Comment (SahilJain314) [inline docs/guides/grpo.md]:
```suggestion
- $\pi_{\theta_{\text{old}}}$ is the previous policy model (from the beginning of this step)
```

File: docs/guides/grpo.md [hunk 7]
```diff
@@ -108,3 +109,56 @@ This Policy object holds a [RayWorkerGroup](../../nemo_reinforcer/distributed/wo
...
+$$
+\begin{align*}
+D_{\text{KL}} (\pi_\theta || \pi_\text{ref}) &\approx E_{x \sim \pi_{\theta}} \Big[ \frac{\pi_\text{ref}(x)}{\pi_\theta(x)} - \log \frac{\pi_\text{ref}(x)}{\pi_\theta(x)} - 1 \Big] \\
+&= \sum_x \pi_{\theta}(x) \Big[ \frac{\pi_\text{ref}(x)}{\pi_\theta(x)} - \log \frac{\pi_\text{ref}(x)}{\pi_\theta(x)} - 1 \Big] \\
+&= \sum_x \pi_{\theta_{\text{old}}}(x) \frac{\pi_{\theta}(x)}{\pi_{\theta_{\text{old}}}(x)} \Big[ \frac{\pi_\text{ref}(x)}{\pi_\theta(x)} - \log \frac{\pi_\text{ref}(x)}{\pi_\theta(x)} - 1 \Big] \\
+&= E_{x \sim \pi_{\theta_\text{old}}} \frac{\pi_{\theta}(x)}{\pi_{\theta_{\text{old}}}(x)} \Big[ \frac{\pi_\text{ref}(x)}{\pi_\theta(x)} - \log \frac{\pi_\text{ref}(x)}{\pi_\theta(x)} - 1 \Big] \\
+\end{align*}
+$$
+
+To enable the on-policy KL approximation, set the config `use_on_policy_kl_approximation=True` in the `ClippedPGLossConfig`.
```
Comment (SahilJain314) [inline docs/guides/grpo.md]:
```suggestion
To enable the on-policy KL approximation, set the config `use_on_policy_kl_approximation=True` in the `ClippedPGLossConfig`. By default, we set this config to False to align with standard GRPO.
```

File: docs/guides/grpo.md [hunk 8]
```diff
@@ -108,3 +109,56 @@ This Policy object holds a [RayWorkerGroup](../../nemo_reinforcer/distributed/wo
...
+
+where:
+
+- $\pi_\theta$ is the policy model we are currently optimizing
+- $\pi_{\theta_{\text{old}}}$ is the previous policy model
+- $A_t$ is the advantage estimate
+- $\varepsilon$ is a clipping hyperparameter
+- $\beta$ is the KL penalty coefficient
+- $\pi_{\text{ref}}$ is the reference policy
+
```
Comment (SahilJain314) [inline docs/guides/grpo.md:128]:
```suggestion

#### Improvements to the GRPO loss formulation for stability and accuracy

```

File: docs/guides/grpo.md [hunk 9]
```diff
@@ -108,3 +109,56 @@ This Policy object holds a [RayWorkerGroup](../../nemo_reinforcer/distributed/wo
...
+\begin{align*}
+E_{x \sim \pi_\text{training}} f_\theta(x) &= \sum_x \pi_\text{training}(x) f_\theta(x) \\
+&= \sum_x \pi_\text{inference}(x) \frac{\pi_\text{training}(x)}{\pi_\text{inference}(x)} f_\theta(x) \\
+&= E_{x \sim \pi_\text{inference}} \frac{\pi_\text{training}(x)}{\pi_\text{inference}(x)} f_\theta(x)
+\end{align*}
+$$
+
+By multiplying the first term of the loss function by the importance weights $\frac{\pi_\text{training}(x)}{\pi_\text{inference}(x)}$, we can correct for the distribution mismatch between $\pi_{\text{training}}$ and $\pi_{\text{inference}}$ while still sampling from $\pi_{\text{inference}}$.
+
+To enable the importance sampling correction, set the config `use_importance_sampling_correction=True` in the `ClippedPGLossConfig`.
```
Comment (SahilJain314) [inline docs/guides/grpo.md]:
```suggestion
To enable the importance sampling correction, set the config `use_importance_sampling_correction=True` in the `ClippedPGLossConfig`. By default, we set this config to False to align with standard GRPO.
```

File: examples/configs/grpo_math_1B.yaml [hunk 10]
```diff
@@ -14,6 +14,8 @@ loss_fn:
   reference_policy_kl_penalty: 0.01
   ratio_eps_min: 0.2
   ratio_eps_max: 0.2
+  use_on_policy_kl_approximation: false
```
Comment (SahilJain314) [inline examples/configs/grpo_math_1B.yaml:18]:
```suggestion
  # (default off) loss formulation improvements (docs/guides/grpo.md#loss)
  use_on_policy_kl_approximation: false
```

File: nemo_reinforcer/algorithms/loss_functions.py [hunk 11]
```diff
@@ -119,9 +125,22 @@ def __call__(
 
         # Calculate KL regularization.
         if self.reference_policy_kl_penalty != 0:
-            kl = self.reference_policy_kl_penalty * calculate_kl_penalty_joschu2020(
-                logprobs_policy=curr_logprobs,
-                logprobs_reference=reference_policy_logprobs,
+            if self.use_on_policy_kl_approximation:
```
Comment (SahilJain314) [inline nemo_reinforcer/algorithms/loss_functions.py:128]:
```suggestion
            if self.use_on_policy_kl_approximation:
                # See: docs/guides/grpo.md#on-policy-kl-approximation
```

File: nemo_reinforcer/algorithms/loss_functions.py [hunk 12]
```diff
@@ -140,7 +159,16 @@ def __call__(
         loss1 = -advantages * ratios
         loss2 = -advantages * ratios_clamped
 
-        actor_loss = masked_mean(torch.max(loss1, loss2), mask)
+        if self.use_importance_sampling_correction:
```
Comment (SahilJain314) [inline nemo_reinforcer/algorithms/loss_functions.py:163]:
```suggestion
        if self.use_importance_sampling_correction:
            # See: docs/guides/grpo.md#importance-sampling-correction
```
